python/ray/serve/controller.py

```diff
@@ -170,22 +168,61 @@ class Config:
 
 
 class BackendState:
-    def __init__(self, checkpoint: bytes = None):
+    def __init__(self,
+                 controller_name: str,
+                 detached: bool,
+                 checkpoint: bytes = None):
+        self.controller_name = controller_name
+        self.detached = detached
+
+        # Non-checkpointed state.
+        self.currently_starting_replicas: Dict[asyncio.Future, Tuple[
+            BackendTag, ReplicaTag, ActorHandle]] = dict()
+        self.currently_stopping_replicas: Dict[asyncio.Future, Tuple[
+            BackendTag, ReplicaTag]] = dict()
+
+        # Checkpointed state.
         self.backends: Dict[BackendTag, BackendInfo] = dict()
+        self.backend_replicas: Dict[BackendTag, Dict[
+            ReplicaTag, ActorHandle]] = defaultdict(dict)
         self.goals: Dict[BackendTag, GoalId] = dict()
+        self.backend_replicas_to_start: Dict[BackendTag, List[
+            ReplicaTag]] = defaultdict(list)
+        self.backend_replicas_to_stop: Dict[BackendTag, List[Tuple[
+            ReplicaTag, Duration]]] = defaultdict(list)
+        self.backends_to_remove: List[BackendTag] = list()
 
         if checkpoint is not None:
-            self.backends, self.goals = pickle.loads(checkpoint)
+            (self.backends, self.backend_replicas, self.goals,
+             self.backend_replicas_to_start, self.backend_replicas_to_stop,
+             self.backend_to_remove) = pickle.loads(checkpoint)
+
+        # Fetch actor handles for all of the backend replicas in the system.
+        # All of these backend_replicas are guaranteed to already exist because
+        # they would not be written to a checkpoint in self.backend_replicas
+        # until they were created.
+        for backend_tag, replica_dict in self.backend_replicas.items():
+            for replica_tag in replica_dict.keys():
+                replica_name = format_actor_name(replica_tag,
+                                                 self.controller_name)
+                self.backend_replicas[backend_tag][
+                    replica_tag] = ray.get_actor(replica_name)
 
     def checkpoint(self):
-        return pickle.dumps([self.backends, self.goals])
+        return pickle.dumps(
+            (self.backends, self.backend_replicas, self.goals,
+             self.backend_replicas_to_start, self.backend_replicas_to_stop,
+             self.backends_to_remove))
 
     def get_backend_configs(self) -> Dict[BackendTag, BackendConfig]:
         return {
             tag: info.backend_config
             for tag, info in self.backends.items()
         }
 
+    def get_replica_handles(self) -> List[ActorHandle]:
+        return self.backend_replicas
+
```

Comment:
Any reason why you are changing the functionality fo this function ?

Reply:
Oops just realized that the type annotation is wrong now -- will fix it. I changed this because similar functionality was used in two places and one of them was accessing a private field. With this change they both use the same function.